src/core/service.c

```diff
@@ -3436,9 +3436,18 @@ static void service_notify_cgroup_empty_event(Unit *u) {
         }
 }
 
-static void service_notify_cgroup_oom_event(Unit *u) {
+static void service_notify_cgroup_oom_event(Unit *u, bool managed_oom) {
         Service *s = SERVICE(u);
 
+        if (managed_oom) {
+                log_unit_debug(u, "Process(es) of control group were killed by systemd-oomd.");
+
+                if (s->exit_type == SERVICE_EXIT_MAIN || cgroup_good(s) <= 0)
+                        s->result = SERVICE_FAILURE_OOM_KILL;
+
+                return;
```

Comment:
I thought the OOMPolicy(continue|stop|kill) was specific to the kernel OOMs and didn't want to mess with it.

So I guess they'll apply to systemd-oomd kills too?

Reply:
Hmm, it shouldn't be applicable really. The kernel will only kill one process at a time, so a service can survive an OOM event. In contrast, systemd-oomd will kill the whole cgroup, which is makes it equivalent to a `OOMPolicy=kill` in most cases.

The exception is if we have `Delegate=yes`, I suppose in that case we will currently *not* see the kill as an OOM kill with this patchset. Do we need something smarter that looks into the leaf nodes if when delegation is on?

Reply:
Ooh. Shloudn't there be *someone else* responsible for checking in delegated subtrees, e.g. it'd be the user instance that manages (user) units in the subtree. (Although, I'm not familiar with the general model of systemd-oomd delegation. Is it delegatable at all?)

Reply:
I doubt it is delegatable really. Right now it pretty much just looks at all leaf nodes and selects the "best" candidate.

That said, systemd-oomd is part of systemd. It could have some logic to (also) set an attribute on the units cgroup rather than only in the child cgroup where the kill happened. Could we search up for the next cgroup with `.service/.scope` in the name and increment `user.oomd_unit_kill`?

Reply:
`OOMPolicy=` shouldn't be applicable to systemd-oomd kills in general as benzea mentioned.

Thinking a bit about delegated subtrees though, we do partially follow what werkov is suggesting: right now the user manager subtrees are delegated and when a systemd-oomd kill happens, the user manager follows the same code paths as what is being implemented in this PR.

Anyways that's just informational. I do like benzea's idea to increment an xattr on the `service / scope` cgroups. Instead of a new xattr though what if we only increment `user.oomd_ooms` on those `service / scope` cgroups? Otherwise it would go unused on the leaf node (unless the leaf is a `service / scope` cgroup). And then in the delegated case if the `service / scope` cgroup is not empty we treat it similar to `OOMPolicy=continue` (that is, the rest of the unit keeps going).

Side note, FB's container workloads are delegated systemd units and have memory.oom.group=1. That way systemd-oomd kills the whole subtree / workload along with any subcgroups created by the workload.

Reply:
I just realized I wrote too much without really giving direction. Looking at what I wrote yesterday and the `service_notify_cgroup_oom_event()` state machine again I have a different opinion. Because `OOMPolicy=kill` sets `memory.oom.group=1` and systemd-oomd respects `memory.oom.group`, I don't see why we shouldn't follow the rest of the case/switch as Lennart mentioned. But we need to update the documentation to say this because right now systemd-oomd kills behave like `OOMPolicy=continue`.

Like benzea mentioned below we have the same problem with kernel OOM kills and memory.events as it doesn't look down the whole subtree. We can make this better for systemd-oomd if we wanted to later on but that's out of the scope of this PR.